src/vivarium_public_health/dataset_manager/artifact.py

```diff
@@ -93,13 +93,17 @@ def write(self, entity_key: str, data: Any):
         ArtifactException :
             If the provided key already exists in the artifact.
         """
+
         entity_key = EntityKey(entity_key)
         if entity_key in self.keys:
             raise ArtifactException(f'{entity_key} already in artifact.')
         elif data is None:
             pass
         else:
             self._keys.append(entity_key)
+            new_keyspace = self._keys.copy()
```

Comment:
I still get a little confused about when you need to copy things in python or not so why are you coping self._keys here and using that instead of just writing self._keys into 'metadata.keyspace'?

Reply:
You only need to copy things if you're hanging on to a mutable data structure (lists, dicts, or sets) and you're handing that data structure to a function that might have side effects.  Python doesn't provide great tools for controlling this behavior like you'd get in a language with a more functional flavor.  

That being said, we shouldn't need to make a copy here.  Functions in the `hdf` module should be side effect free w/r/t their input arguments.  If that's not the case, we should look there to make the changes (e.g. have those functions copy their args before mutating).

Reply:
Actually it was wrong to begin with. Since we all forgot about the fact that we decided to have metadata as strings, not `EntityKey` means, `new_keyspace=[str(k) for k in self._keys]` and yes, no copy